### DISCUSSION TITLE
[Idea/Fix] Improve permalink copy and paste

### DISCUSSION THREAD

**Discussion by AnthonyMuscio** [TiddlyWiki/TiddlyWiki5#5935] [Ideas]
Folks,

**Problem**
Please see my Post in the forum [Notes from the Coalface: Pasting URL's with pretty title](https://groups.google.com/g/tiddlywiki/c/F-dQxKm4rds) when we use the permalink to copy to the clipboard it is only passing the raw URL with its encoding. Thus when we paste it elsewhere is loses it's pretty title. This includes when we past anywhere else.

As my post points out "this need not be the case", there is clearly a way to copy it into the clipboard (In windows at least)  such as pasting into rich text type editors, the pretty link is retained, while copying into a plain text field as in tiddlywiki is remains an ugly, though working link.

**My suggestions**
- First I would like to suggest we find a way to copy permalinks to the clipboard while retaining this duality of encoded URL/and or URL+prity link.
- Secondly it would be nice to have a way that when pasting such an improved link into tiddlywiki that we could get it to paste either as HTML with pretty link OR using the tiddlywiki form [[pretty link|https://url]] format

This would simplify even further the drag and drop or copy and paste of tiddlywiki permalinks or any other link into tiddlywiki perhaps even allowing us to bypass some of the bookmarking tools developed recently that are in part a response to this limitation. 

**Design notes;**
Perhaps for backward compatibility we create a new permalink button, and if necessary provide drop or paste area that converts such links to tiddlywiki (or markdown) formats.

Tony

**Comment by AnthonyMuscio** [comment]
It would be nice if we could actually interrogate and use the real content of the clipboard using "deserializers" on the clipboard content for more interactive solutions. Perhaps this needs to be a browser add on?

**Reply by AnthonyMuscio** [reply]
In firefox you can r-click and inspect, choose the html link and copy the outer html result of the subsequent paste is for example;

`<a class="tc-tiddlylink tc-tiddlylink-resolves" href="#Anchor%20Links%20using%20HTML"><h2 class="tc-title">Anchor Links using HTML</h2></a>`

<a class="tc-tiddlylink tc-tiddlylink-resolves" href="#Anchor%20Links%20using%20HTML"><h2 class="tc-title">Anchor Links using HTML</h2></a>

It not only brings the pretty link but the CSS classes.

However it would be better if we could do this or something similar with permalink button. Perhaps Permalink as html?

Regards
Tony

**Reply by pmario** [reply]
I think, it depends, where you want to paste the link. The format we have to create will change. Here at github it would be a markdown syntax. If we use HTML syntax here, it creates an XL link, which I personally consider aggressive. 

For the GG it will understand the A element. 

For TW we will need the prettylink syntax. 

The only problem I see is TW-permaview. ... We will need a dialogue, where we can enter the alt-text, since permaview contains several tiddlers.

We may experiment with 2 new buttons, that display a modal, that will allow us to create the different link types. So it will be a 2 or 3 click UX instead of a 1 click.

**Reply by AnthonyMuscio** [reply]
Mario,

Thanks for reading this. Sorry, my follow-up reply was just peaking inside links by showing the inspect link copy method. I am not saying this is necessarily the way to go.

In my inspect and copy example above it captured the H2 in the link title.

Perhaps my research on this needs to go further.

If I may restate,

On windows I can highlight and copy a tiddlywiki title/link and past it into Word, Outlook, google groups, gmail and it becomes an active link with the pretty name and html tags. With the same copy I can paste it into GitHub, Tiddlywiki text areas as plain text only (with the link info forgotten). **Of interest is in Gmail and GG if you paste into the subject its pasted as plain text, if you paste into the body of the email or post, it becomes a pretty link.** 

That is the clipboard actually contains both the link and the pretty link, but pasting into a text area reduces it to text only and as a result you need to use the permalink button, to copy the link URL as plain text for pasting into tiddlywiki or git hub then in a separate action copy the pretty link and form a mark-up link.

Speculation

I am wondering if there is a way to get tiddlywiki to accept the paste and retaining the link details within it? Even if we need a different area 
- if possible reformatting to use tiddlywiki syntax
- alternatively a html a tag.
- or have a permalink copy work like just like the highlight and copy

Possibilities

If we accept a link paste, say as an a tag/href into a paste area/field from the clipboard, we can easily build a method to parse this into a tiddlywiki format link. eg extract the href value and its pretty link.

The technical gaps
- Getting tiddlywiki to behave like google products to accepting a past of a link and preserve the pretty link
- Getting tiddlywikis copy to clipboard or a new permalink copy to the windows clipboard, to behave the same way highlight and copy does.

Perhaps the way I could start is to make copy to clipboard buttons copy to different link formats but this still does not address the paste from elsewhere.

Regards
Tones

**Reply by Jermolene** [reply]
Hi @AnthonyMuscio at the moment the core only supports copying text strings to the clipboard. We'd need a new primitive to be able to copy HTML strings to the clipboard; it would be reasonable to add it, I think, for many applications.

**Reply by AnthonyMuscio** [reply]
Yes that would be great. I expect once the capacity is there we can develop tools on top of it.
